core/src/main/java/org/apache/calcite/rel/rules/MinusToDistinctRule.java

```diff
@@ -0,0 +1,184 @@
+
+import java.math.BigDecimal;
+
+/**
+ * Planner rule that translates a distinct
+ * {@link org.apache.calcite.rel.core.Minus}
+ * (<code>all</code> = <code>false</code>)
+ * into a group of operators composed of
+ * {@link org.apache.calcite.rel.core.Union},
+ * {@link org.apache.calcite.rel.core.Aggregate},
+ * {@link org.apache.calcite.rel.core.Filter},etc.
+ *
+ * <p>For example, the query plan
+
+ * <blockquote><pre>{@code
+ *  LogicalMinus(all=[false])
+ *    LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2])
+ *      LogicalFilter(condition=[=($7, 10)])
+ *        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+ *    LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2])
+ *      LogicalFilter(condition=[=($7, 20)])
+ *        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+ * }</pre></blockquote>
+ *
+ * <p> will convert to
+ *
+ * <blockquote><pre>{@code
+ *  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2])
+ *    LogicalFilter(condition=[AND(>($3, 0), =($4, 0))])
+ *      LogicalAggregate(group=[{0, 1, 2}], agg#0=[COUNT() FILTER $3], agg#1=[COUNT() FILTER $4])
+ *        LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], $f3=[=($3, 0)], $f4=[=($3, 1)])
+ *          LogicalUnion(all=[true])
+ *            LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], $f3=[0])
+ *              LogicalFilter(condition=[=($7, 10)])
+ *                LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+ *            LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], $f3=[1])
+ *              LogicalFilter(condition=[=($7, 20)])
+ *                LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+ * }</pre></blockquote>
+ *
+ * @see CoreRules#MINUS_TO_DISTINCT
+ */
+@Value.Enclosing
+public class MinusToDistinctRule
+    extends RelRule<MinusToDistinctRule.Config>
+    implements TransformationRule {
+
+  protected MinusToDistinctRule(Config config) {
+    super(config);
+  }
+
+  @Deprecated // to be removed before 2.0
+  public MinusToDistinctRule(Class<? extends Minus> minusClass,
+      RelBuilderFactory relBuilderFactory) {
+    this(MinusToDistinctRule.Config.DEFAULT.withRelBuilderFactory(relBuilderFactory)
+        .as(MinusToDistinctRule.Config.class)
+        .withOperandFor(minusClass));
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Minus minus = call.rel(0);
+
+    if (minus.all) {
+      // Nothing we can do
+      return;
+    }
+
+    final RelOptCluster cluster = minus.getCluster();
+    final RelBuilder relBuilder = call.builder();
+    final RexBuilder rexBuilder = cluster.getRexBuilder();
+    final int branchCount = minus.getInputs().size();
+
+    // For each child branch in minus,add a column which indicates branch index
+    //
+    // e.g. select EMPNO from emp -> select EMPNO,0 from emp
+    // 0 indicates that it comes from minus the first child branch
+    for (int i = 0; i < branchCount; i++) {
+      relBuilder.push(minus.getInput(i));
+      relBuilder.projectPlus(relBuilder.literal(new BigDecimal(i)));
```

Comment:
Nit: since this is a counter for the branch numbers, maybe `BigDecimal` is an overkill?